cmd/cosign/cli/options/certificate.go

```diff
@@ -75,12 +77,25 @@ func (o *CertVerifyOptions) AddFlags(cmd *cobra.Command) {
 	cmd.Flags().StringVar(&o.CertGithubWorkflowRef, "certificate-github-workflow-ref", "",
 		"contains the ref claim from the GitHub OIDC Identity token that contains the git ref that the workflow run was based upon.")
 	// -- Cert extensions end --
+	cmd.Flags().StringVar(&o.CAIntermediates, "ca-intermediates", "",
+		"path to a file of intermediate CA certificates in PEM format which will be needed "+
+			"when building the certificate chains for the signing certificate. "+
+			"The flag is optional and must be used together with --ca-roots, conflicts with "+
+			"--certificate-chain.")
+	_ = cmd.Flags().SetAnnotation("ca-intermediates", cobra.BashCompFilenameExt, []string{"cert"})
+	cmd.Flags().StringVar(&o.CARoots, "ca-roots", "",
+		"path to a bundle file of CA certificates in PEM format which will be needed "+
+			"when building the certificate chains for the signing certificate. Conflicts with --certificate-chain.")
+	_ = cmd.Flags().SetAnnotation("ca-roots", cobra.BashCompFilenameExt, []string{"cert"})
+
```

Comment:
I'm less familiar with cosign so please take this with a grain of salt, but: I'm personally -1 on separate state flags for intermediate and root CA certificates.

My rationale:

1. Technically speaking, there is no real distinction between a "root" and an "intermediate" certificate: some people use "root" to mean "self-signed," but it's common to have both non-self-signed and cross-issued roots in real PKIs. In other contexts "intermediate" means "part of the untrusted set", but it's not immediately clear to me whether intermediates are trusted or untrusted in cosign (CC @haydentherapper)
2. Long term, I believe the plan is to have most Sigstore clients use a common `--trusted-root` or similar flag, which will load a shared format for the roots of trust for an "entire" Sigstore deployment (including the Rekor key, CTFE key, etc.). While this doesn't preclude that effort, it does add another set of knobs that will eventually need to be deprecated (or treated as another source of state). 
3. Similarly long-term, I believe the plan is to have most Sigstore clients (including `cosign`) use Sigstore bundles as their interchange format. Bundles in turn support untrusted intermediates for the BYO PKI case, and in an unambiguous fashion (everything in the bundle is presumed untrusted). 

So, TL;DR: if the idea behind `--ca-intermediates` is for it to pass *trusted* intermediates, then IMO it should be removed and collapsed into `--ca-roots` (which IMO could be called `--trusted-certs` to minimize confusion).

On the other hand, if these certificates are meant to be *untrusted* I can see the value of this flag, but I think it might be somewhat short-lived given the trustroot + bundle progression. But maybe that isn't as far along as I think 😅

Reply:
For public good Sigstore, the intermediate is "trusted", as it is provided via the TUF trust root, and it's part of the [trusted_root.json](https://github.com/sigstore/root-signing/blob/main/repository/repository/targets/cec894ad77f79b1cb324150f6363012bcef7492954f3ab9134f932e6aa2e2e20.trusted_root.json#L50), and for public good we do not allow intermediates not provided direct via the trust root.

But if `cosign` is to be used with other deployments, they can be untrusted. So with that in mind, and the bundle format is not yet supported in cosign (to carry untrusted intermediates) I think the current proposal makes sense, but it can feel a bit awkward for the public good instance.

Reply:
@woodruffw thanks for your review and suggestions! 

> Technically speaking, there is no real distinction between a "root" and an "intermediate" certificate

Well, for example looking at the Fulcio [Certificate Specification](https://github.com/sigstore/fulcio/blob/main/docs/certificate-specification.md), it has subsections for [Root](https://github.com/sigstore/fulcio/blob/main/docs/certificate-specification.md#root-certificate), [Intermediate](https://github.com/sigstore/fulcio/blob/main/docs/certificate-specification.md#intermediate-certificate), and [Issued](https://github.com/sigstore/fulcio/blob/main/docs/certificate-specification.md#issued-certificate) Certificates, with specifications of what each type of certificate `MUST` / `MUST NOT` / `SHOULD`  have. Also, for better or worth, a search for `intermediate` brings up over 140 hits in the `sigstore/cosign` codebase, including in the existing options for `cosign verify`:

> doc/cosign_verify.md:      --certificate-chain string                                                                 path to a list of CA certificates in PEM format which will be needed when building the certificate chain for the signing certificate. Must start with the parent intermediate CA certificate of the signing certificate and end with the root certificate

also over 60 matches have both `intermediate` and `root` in the same line so these terms seems both well established and useful in the `cosign` repo and code.  Maybe it is due to my 'surrounding environment' but I don't remember any cases where I or people around would have difficulties distinguishing which (CA) certificates are  root ones, and which - intermediates.

> In other contexts "intermediate" means "part of the untrusted set", but it's not immediately clear to me whether intermediates are trusted or untrusted in cosign
> [...]
> But if cosign is to be used with other deployments, they can be untrusted. 

I have to admit having difficulties understanding which certificates are (or should be) considered "trusted" and "untrusted" in cosign (there is only one search hit on `untrusted` in sigstore/cosign and it is to a CHANGELOG in https://github.com/sigstore/cosign/pull/2124; the only [search hit](https://docs.sigstore.dev/about/threat-model/) on `untrusted` in https://docs.sigstore.dev also is not very helpful).  I think the certificate trust is so much context-dependant.  For example, I'm used to think of the certificates issues by our `TheCompanyCA`  as both fully trusted and more trustworthy than those from Digicert & Co. 😄  - while in an environment oriented to the public websites and Certificate Authorities it would likely be considered as "Self-Signed" and untrusted.

Regarding the points 2. and 3.  ("use a common --trusted-root flag" and Sigstore bundles as the interchange format) - in the long term, I have some reservations regarding this in terms of the ease of use and Developer / DevOps Experience.  I worry that they would become custom requirements and make hurdles to using Sigstore for signing artifacts.  PEM certificate bundles are "universal", ubiquitous and extremely familiar to everyone working with them and also deeply interwined with the existing scripts, processes etc.  The TUF roots / Sigstore bundles are (IMO) in comparison relatively new, more complex (PEM -> JSON -> ...),  "special purpose" and require custom handling, adoption steps / timeframe and increase the adoption and operational friction, especially for large users.  Undoubtedly, the change will make life easier for us as the sigstore developers and the codebase probably nicer ("_just one file/format to rule them all!_") but possibly at the expense of placing some extra burden on the end users ("_in addition to distributing the `ca-bundle.pem` to 100,000 endpoints that you are doing now and must continue to do for many years, you also need to create the <new format> file and distribute it to all the endpoints that will or might need to verify cosign signatures_").   So I would recommend and plead 😄  to consider the trade-off carefully and if decided, let the users know well in advance about the upcoming change(s) and provide for an ample transition period (maybe something like one major version during which one could use both _old_ and _new_ ways of supplying certificates) 🙏 /cc @haydentherapper 

But as you said, this would be long-term changes - and currently we have an "acute" issue of not being able to (properly) verify cosign signatures with `cosign verify` and through the [cosign Go library](https://pkg.go.dev/github.com/sigstore/cosign) if there is more than one CA authority issuing certificates - namely  using a certificate bundle PEM file that contains  multiple CA Root certificates.  (And we would like to start issuing and verifying _lots_ of Sigstore/Cosign signatures **in Q1 '24**! 😁).

The suggested `cosign verify --ca-roots <certbundle.pem>` and the expanded [CertVerifyOptions](https://pkg.go.dev/github.com/sigstore/cosign/cmd/cosign/cli/options#CertVerifyOptions) and [VerifyCommand](https://pkg.go.dev/github.com/sigstore/cosign/cmd/cosign/cli/verify#VerifyCommand) in the Go library would solve that problem.  If you have other ideas how we could do this (in the near term), please let me know!  (I thought about expanding the current `--certificate-chain` format or adding `--certificate-chains` [plural], but couldn't figure out how to handle the file format - with the single chain, it is easy to parse the file [`<Intermediate>*<Root>`] but if there are multiple roots, it gets very messy....  I also didn't want to get into the business of building multiple certificate chains from a set of certificates - the Go stdlib already does it very well but it requires providing `Roots` and optionally `Intermediates` in the [VerifyOptions](https://pkg.go.dev/crypto/x509#VerifyOptions) mentioned above - to which the two new suggested flags would correspond.)